JIT: Optimize range checks for X >> CNS and for short/byte indices

EgorBo opened the pull request:
This PR optimizes range checks for the following cases:
```csharp
class Tests
{
    // Case 1: index is char => it's never negative
    void Test1(int[] array, char i)
    {
        if (i < array.Length)
            array[i] = 0;
    }

    // Case 2: index is byte => it's never negative
    char Test2(int[] array, byte i)
    {
        if (i < 4)
            return "abcd"[i];

        return ' ';
    }

    // Case 3: X >> CNS
    char Test3(char i)
    {
        if (i < 128)
            return "abcdefgh"[i >> 4];
        return '\0';
    }

    // Case 4: Array has >= index.GetType().MaxValue elements
    static ReadOnlySpan<byte> s_hexDecodeMap => new byte[256] {..}; // 256 elements

    byte Test4(byte index) => s_hexDecodeMap[index]; // never goes out of bounds
}
```

E.g.:
```csharp
char Test(int i)
{
    if ((uint)i < 26)
        return "abcdefghijklmnopqrstuvwxyz"[i >> 4];
    return '\0';
}
```
```diff
; Method Proga:Test(int):ushort
-G_M46674_IG01:              
-      4883EC28             sub      rsp, 40
G_M46674_IG02:              
       83F91A               cmp      ecx, 26
       7321                 jae      SHORT G_M46674_IG05
G_M46674_IG03:              
       C1F904               sar      ecx, 4
-      83F91A               cmp      ecx, 26
-      7320                 jae      SHORT G_M46674_IG07
       48B8484400DD8B020000 mov      rax, 0x28BDD004448
       488B00               mov      rax, gword ptr [rax]
       8BD1                 mov      edx, ecx
       0FB744500C           movzx    rax, word  ptr [rax+2*rdx+12]
G_M46674_IG04:              
-      4883C428             add      rsp, 40
       C3                   ret      
G_M46674_IG05:              
       33C0                 xor      eax, eax
G_M46674_IG06:              
-      4883C428             add      rsp, 40
       C3                   ret      
-G_M46674_IG07:              
-      E8EA0A555F           call     CORINFO_HELP_RNGCHKFAIL
-      CC                   int3     
; Total bytes of code: 55
```
@stephentoub said this pattern is used for Regex

let's see if there are any diffs

EgorBo committed:
Handle X >> CNS rangecheck

EgorBo committed:
Fix comments

stephentoub commented:
> @stephentoub said this pattern is used for Regex

Thanks. The pattern used by regex is actually:
```
char Test(char i)
{
    if (i < 128) // where the const is 128 or less for an 8 char string
        return "abcdefgh"[i >> 4];
    return '\0';
}
```

EgorBo commented:
> > @stephentoub said this pattern is used for Regex
> 
> Thanks. The pattern used by regex is actually:
> 
> ```
> char Test(char i)
> {
>     if (i < 128) // where the const is 128 or less for an 8 char string
>         return "abcdefgh"[i >> 4];
>     return '\0';
> }
> ```

ah, then I need to properly calculate the range:
x is `[0..128]`, x >> 4 is `[0>>4..128>>4]`

EgorBo commented:
@stephentoub is that worth doing or https://github.com/dotnet/runtime/pull/67133 will handle it on the Regex side?

stephentoub commented:
> @stephentoub is that worth doing or https://github.com/dotnet/runtime/pull/67133 will handle it on the Regex side?

It's worth doing if it's not too hard.  67133 doesn't change the bounds check, it just reduces in some situations how often we need to lookup in the bitmap.  If you have a character class in a regex like [A-Za-z], 67133 just makes it so that rather than doing:
```C#
if (c < 128 && stringWith8Chars[c >> 4] ....
```
we'll do:
```C#
if (c < 123 && stringWith8Chars[c >> 4] ...
```
such that we don't have to do the lookup for a handful of chars.

stephentoub commented:
> let's see if there are any diffs

This pattern will show up in the majority of regexes, but in the code generated by RegexCompiler (via DynamicMethod) and the source generator, so it likely won't hit in regex code in our diffs.

EgorBo committed:
Properly compute ranges

EgorBo committed:
Fix RangeCheck::DoesOverflow

EgorBo commented:
Fixed
![image](https://user-images.githubusercontent.com/523221/160177051-c3f6e33e-7e96-4c7d-bf61-3bedcf8e4832.png)


as a side effect I fixed bound checks for `char` and `byte` index, e.g.:
```csharp
void Test1(int[] array, char i)
{
    if (i < array.Length)
        array[i] = 0;
}
```

EgorBo commented:
So this fixes https://github.com/dotnet/runtime/issues/13159

stephentoub commented:
Awesome!!

EgorBo committed:
fix assert

EgorBo committed:
Clean up

EgorBo committed:
Clean up

EgorBo commented:
PTAL @SingleAccretion @anthonycanino who knows/contributed to this area 🙂

EgorBo committed:
Clean up, fix build

EgorBo commented:
@dotnet/jit-contrib PTAL

SingleAccretion commented on src/coreclr/jit/rangecheck.cpp:
```diff
@@ -1261,12 +1268,57 @@ bool RangeCheck::ComputeDoesOverflow(BasicBlock* block, GenTree* expr)
     {
         overflows = DoesPhiOverflow(block, expr);
     }
+    else if (expr->OperIs(GT_CAST) && expr->AsCast()->CastOp()->OperIs(GT_LCL_VAR))
+    {
+        // See if this is a cast on top of an argument - in this case we won't ever overflow
+        var_types      castType = expr->AsCast()->CastToType();
+        GenTreeLclVar* lcl      = expr->gtGetOp1()->AsLclVar();
+        if ((lcl->GetSsaNum() == SsaConfig::FIRST_SSA_NUM)) // make sure it's the first def of the arg
+        {
+            LclVarDsc* varDsc = m_pCompiler->lvaGetDesc(lcl->GetLclNum());
+            if (varDsc->lvIsParam && (varDsc->TypeGet() == castType) && varTypeIsSmall(castType))
+            {
+                overflows = false;
+            }
+        }
+    }
```
When does a cast "overflow" by this definition of "overflow" in the general case?

EgorBo replied:
no idea, the whole function DoesOverflow which is able to "cancel" bound check elimination is questionable, but I don't want to remove as part of this PR

EgorBo committed:
Update src/coreclr/jit/rangecheck.cpp

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

EgorBo committed:
Merge branch 'main' of https://github.com/dotnet/runtime into bound-checks-rsh

EgorBo committed:
Merge branch 'main' of https://github.com/dotnet/runtime into bound-checks-rsh

EgorBo committed:
address feedback

EgorBo commented on src/coreclr/jit/rangecheck.cpp:
```diff
@@ -1261,12 +1268,57 @@ bool RangeCheck::ComputeDoesOverflow(BasicBlock* block, GenTree* expr)
+        case TYP_SHORT:
+            return Range(Limit(Limit::keConstant, INT16_MIN), Limit(Limit::keConstant, INT16_MAX));
+        case TYP_UINT:
+            return Range(Limit(Limit::keConstant, 0), Limit(Limit::keConstant, UINT32_MAX));
```
Thanks, decided to remove i32 cases for now

EgorBo commented on src/coreclr/jit/rangecheck.cpp:
```diff
@@ -1372,31 +1424,32 @@ Range RangeCheck::ComputeRange(BasicBlock* block, GenTree* expr, bool monIncreas
+        range = Range(Limit(Limit::keUnknown));
+
+        // See if this is a cast on top of an argument - in this case we won't ever overflow
+        // and can, at least, use castType's range
```
slightly changed the code and left a TODO to compute intersection of cast type and castop

EgorBo commented:
@SingleAccretion anything else catches your eye? 🙂

SingleAccretion commented on src/coreclr/jit/rangecheck.cpp:
```diff
@@ -1033,6 +1067,11 @@ Range RangeCheck::ComputeRangeForLocalDef(BasicBlock*          block,
     LclSsaVarDsc* ssaDef = GetSsaDefAsg(lcl);
     if (ssaDef == nullptr)
     {
+        LclVarDsc* varDsc = m_pCompiler->lvaGetDesc(lcl->GetLclNum());
+        if (varDsc->lvIsParam && (lcl->GetSsaNum() == SsaConfig::FIRST_SSA_NUM))
+        {
+            return GetRangeFromType(varDsc->TypeGet());
+        }
```
Heh, parameters are `NormalizeOnLoad`, so the small type for them doesn't mean anything (they can have any `TYP_INT` value, unless normalized by a preceding store, but here the check is very specifically for the out-of-method definition).

Is it actually needed?

EgorBo committed:
Merge branch 'main' of https://github.com/dotnet/runtime into bound-checks-rsh

EgorBo committed:
address feedback

EgorBo committed:
Merge branch 'main' of https://github.com/dotnet/runtime into bound-checks-rsh

JulieLeeMSFT commented:
@dotnet/jit-contrib PTAL.

anthonycanino reviewed (approved):
The changes LGTM after reading @SingleAccretion comments and going through the code myself.

I have one high level comment, but that is not a blocker to the PR. I encoded LSH as a transformation to a multiply, as it let use one multiply implementation to cover the two cases (MUL, LSH). Can we do something similar for RSH, with an (integer) divide, and get two cases (DIV, RSH)?

EgorBo commented:
@BruceForstall PTAL

BruceForstall commented on src/coreclr/jit/rangecheck.h:
```diff
@@ -287,6 +307,21 @@ struct RangeOps
         }
     }
 
+    // Given a constant limit in "l1", do ">>" it to l2 and mutate "l2".
```
This comment is confusing (I realize you mostly cut-and-paste it from above). For one thing, `l2` never is mutated (maybe the arguments should be `const Limit&`). Also, "do '>>' it to l2" doesn't make sense. And it seems weird that the operation is basically `l2 >> l1` instead of `l1 >> l2`.

EgorBo replied:
Good point, addressed

BruceForstall commented on src/coreclr/jit/rangecheck.cpp:
```diff
@@ -1024,9 +1024,43 @@ Range RangeCheck::ComputeRangeForBinOp(BasicBlock* block, GenTreeOp* binop, bool
+// Return value:
+//   range that represents the values given type allows
+//
+static Range GetRangeFromType(var_types type)
```
Please make this a static function of `RangeCheck` instead of a "loose" function.

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into bound-checks-rsh

EgorBo committed:
Address Bruce's feedback
